[FLINK-26432] Cleanly separate validator, observer and reconciler modules

gyfora opened the pull request:
This ticket solves the problem of the lingering configmaps after deleting HA clusters by trying to gracefully shutdown jobs before deleting everything.

It makes the deletion heavier and prone to timeoutexceptions , but this could be improved if we merge https://github.com/apache/flink-kubernetes-operator/pull/21 first then I can improve this based on that.

Also async operations would make this nicer but that is a separate effort tracked in a different ticket.

gyfora commented:
cc @tweise @wangyang0918

tweise commented:
Looks good, will review after it is rebased since there is a larger conflict now.

gyfora commented:
@tweise I rebased the changes to the PR and in order to be able to reuse the functionality you added to the reconciler I had to make some minor changes to the deployment checking.

The logic for reconciliation stays the same but the deployment status checking is more explicit and reusable in other parts now.

gyfora commented:
cc @wangyang0918 @morhidi

wangyang0918 reviewed (commented):
@gyfora Thanks for creating this PR. I left few comments and some of them might be out the scope of this PR.

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/BaseReconciler.java:
```diff
@@ -75,29 +83,44 @@ public boolean removeDeployment(FlinkDeployment flinkApp) {
             }
         }
-        return null;
+        return JobDeploymentStatus.MISSING;
```
This is indeed confusing. It previously meant that the deployment is ready and subsequent checks can proceed.

tweise replied:
According to that it should be `return JobDeploymentStatus.READY`?

gyfora replied:
this is a mistake indeed, fixing

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/BaseReconciler.java:
```diff
@@ -54,11 +65,8 @@ public boolean removeDeployment(FlinkDeployment flinkApp) {
-            FlinkService flinkService) {
+    protected JobDeploymentStatus checkJobManagerDeployment(
+            FlinkDeployment flinkApp, Context context, Configuration effectiveConfig) {
         if (!jobManagerDeployments.contains(flinkApp.getMetadata().getUid())) {
```
Let's keep that out of this PR. However, I don't think we would want to repeat port ready check on every reconciliation, even if deployment is cached. I will think a bit more about it when implementing the terminal error state for https://issues.apache.org/jira/browse/FLINK-26261

wangyang0918 replied:
Yes. We indeed do not want to repeat port ready check on every reconciliation. It is a heave operation. Looking forward the improvement in FLINK-26261.

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobDeploymentStatus.java:
```diff
@@ -0,0 +1,50 @@
+
+/** Status of the Flink job deployment. */
+public enum JobDeploymentStatus {
+    READY,
```
Document what these mean?

gyfora replied:
makes sense, will add some comments to the values

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/BaseReconciler.java:
```diff
@@ -75,29 +83,44 @@ public boolean removeDeployment(FlinkDeployment flinkApp) {
-                            // reconciliation
-                            return null;
-                        }
+                        return JobDeploymentStatus.READY;
```
The above check was required to delay job check to not run into timeout while skipping this for pre-existing deployment. How is that covered now?

gyfora replied:
JobDeploymentStatus.READY means that the deployment and ports are ready. In that case there should not be any timeout. Did I overlook something?

tweise replied:
There is a period of typically a few seconds between port ready and rest server ready and if you hit that then the list jobs will run into a timeout. And hence it is also necessary to make a distinction for existing deployments and operator restart, where that isn't relevant. I think I will start covering these scenarios with test cases since we cannot afford to continue moving things around w/o regression checks.

gyfora replied:
I will try to make sure this case is accounted for with a test case.

gyfora commented:
Thank you @tweise and @wangyang0918 , I will address the comments in the original commit.

I now pushed an optional refactor commit that restores the observe/reconcile flow that we had before @tweise 's recent change while making it clearer and including the jobmanager deployment validation. 

I would like to hear your opinion about this. I think this makes the reconciler simpler and the flow cleaner but this is an optional change if you feel otherwise.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -86,16 +85,14 @@ public FlinkDeploymentController(
-                kubernetesClient,
-                true);
+        LOG.info("Stopping cluster {}", flinkApp.getMetadata().getName());
         getReconciler(flinkApp).removeDeployment(flinkApp);
```
good catch

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobReconciler.java:
```diff
@@ -74,11 +72,14 @@ public JobReconciler(KubernetesClient kubernetesClient, FlinkService flinkServic
                     Optional.ofNullable(jobSpec.getInitialSavepointPath()));
             IngressUtils.updateIngressRules(
                     flinkApp, effectiveConfig, operatorNamespace, kubernetesClient, false);
+            return UpdateControl.updateStatus(flinkApp)
```
makes it clearer, I agree

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobReconciler.java:
```diff
@@ -179,4 +180,22 @@ private void printCancelLogs(UpgradeMode upgradeMode, String name) {
+            }
+        }
+
+        FlinkUtils.deleteCluster(flinkApp, kubernetesClient, true);
```
Do you mean we should delete it always?

wangyang0918 replied:
My bad. I miss the return in the `try {...}`.

tweise commented:
@gyfora I'm having a hard time following what's going on here, similar to how hard it was to change to the original reconciliation logic. Can we take a step back and discuss how we want this to work before any further code mods? I think after that we will also need to add more test coverage to assert the expected behavior.

gyfora commented:
I agree this would be a good time to discuss and agree on a few design principles. 

In my view the Validate -> Observe -> Reconcile flow is very natural and easy to understand.  We want to keep each of these components simple and self contained as much as possible. 

The commits that introduced jobmanager deployment validation (which is a very important, core feature) removed the original observer -> reconcile logic and moved part of the observing to the reconciler. To me this made the whole reconcile very complex and hard to understand with null checks and logic duplication.

I think it leads to a simpler design to have a separation of concerns where the Observer component can describe the status of the deployment and the job, and the reconciler can act on it.

If there is disagreement here we should probably move this to the mailing list for further discussion and visibility.

gyfora commented:
For the sake of transparency I have moved this to a discussion thread on the dev list: https://lists.apache.org/thread/ydvdxfn8go95gnmmh5k9ggwx9hwxn6tl

Let's try to come to an agreement there together before continuing! :)

wangyang0918 commented:
Maybe we could get this PR continued and create a dedicated ticket for refactoring/rethinking the controller flow since we need some time to collect more feedbacks from the ML.

gyfora commented:
> Maybe we could get this PR continued and create a dedicated ticket for refactoring/rethinking the controller flow since we need some time to collect more feedbacks from the ML.

Yea, I will keep this PR open and rename it to dedicate to the refactor discussion. And I will move the first commit to a new PR (once I addressed all the comments) which is related to the original ticket

gyfora commented on helm/flink-operator/crds/flinkdeployments.flink.apache.org-v1.yml:
```diff
@@ -9073,6 +9073,13 @@ spec:
                     type: string
                 type: object
+              jobManagerDeploymentStatus:
+                enum:
```
@tweise you will probably need to add a special status for deployment errors. I have not added a state for that as we have no logic in place yet.

Other than that we can probably reuse the reconciliation status error field for the actual error message.

gyfora commented:
@wangyang0918 @morhidi @Aitozi @bgeng777 
If you guys have time to review this today, I would  be happy to merge it at the end of the day

SteNicholas commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java:
```diff
@@ -0,0 +1,195 @@
+                    return;
+                } else {
+                    // typically it takes a few seconds for the REST server to be ready
+                    LOG.info(
```
Does this need to wait for the port ready? IMO, this could wait for the REST server ready and change the `JobManagerDeploymentStatus` to `DEPLOYED_NOT_READY `.

SteNicholas commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/BaseReconciler.java:
```diff
@@ -54,81 +47,22 @@ public BaseReconciler(KubernetesClient kubernetesClient, FlinkService flinkServi
-                == JobManagerDeploymentStatus.READY) {
+        if (JobManagerDeploymentStatus.READY
+                == flinkApp.getStatus().getJobManagerDeploymentStatus()) {
             shutdown(flinkApp, effectiveConfig);
```
Does this need to delete cluster after shutdown?

gyfora replied:
No, the shutdown operation cleans up the resources (we should document this more in the code), so only need to delete the ingress afterwards

SteNicholas reviewed (commented):
@gyfora, I left the minor comments for the changes.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/JobManagerDeploymentStatus.java:
```diff
@@ -44,12 +44,12 @@
                 return UpdateControl.updateStatus(flinkDeployment)
                         .rescheduleAfter(PORT_READY_DELAY_SECONDS, TimeUnit.SECONDS);
-            case READY:
             case MISSING:
```
Which branch are you suggesting to be removed? the "MISSING"?

SteNicholas replied:
@gyfora Yes.

gyfora replied:
I would prefer to keep it just for transparency, technically it's not needed but I think it makes it cleaerer

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java:
```diff
@@ -0,0 +1,195 @@
+                    return;
+                } else {
+                    // typically it takes a few seconds for the REST server to be ready
+                    LOG.info(
```
Maybe it would be better to log this when the port became ready, I agree :)

SteNicholas replied:
@gyfora, this makes sense to me.

wangyang0918 reviewed (commented):
Thanks for updating this PR. I left some comments and will do a manual test.

wangyang0918 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobReconciler.java:
```diff
@@ -174,15 +159,17 @@ private void printCancelLogs(UpgradeMode upgradeMode, String name) {
     protected void shutdown(FlinkDeployment flinkApp, Configuration effectiveConfig) {
-        if (flinkApp.getStatus().getJobStatus() != null
-                && flinkApp.getStatus().getJobStatus().getJobId() != null) {
+        if (org.apache.flink.api.common.JobStatus.RUNNING
```
Maybe we do not need to ignore case.

gyfora replied:
I will add ignoreCase just to be safe and this needs to be hardened

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java:
```diff
@@ -0,0 +1,183 @@
+        return isReadyToReconcile(flinkApp, effectiveConfig);
+    }
+
+    private void observeJmDeployment(
```
Yes this is true, we probably need a FAILED state to handle some of these cases and some further changes in other places. Better leave this to a follow up I agree

wangyang0918 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java:
```diff
@@ -0,0 +1,182 @@
+                            "JobManager deployment {} in namespace {} port ready, waiting for the REST API...",
+                            flinkApp.getMetadata().getName(),
+                            flinkApp.getMetadata().getNamespace());
+                    deploymentStatus.setJobManagerDeploymentStatus(
```
What will happen if all the replicas are running, but the rest port is not reachable? We will have the `MISSING`, not the `DEPLOYED_NOT_READY`.

gyfora replied:
Exactly, thats how it works and this is also how it worked before, it stays in DEPLOYING until the replicas and ports are ready when it goes to DEPLOYED_NOT_READY. The name might not be the best but it means that we need to do one more iteration because the rest api probably need a few more seconds (which we cannot explicitly check)

gyfora replied:
This is now highlighted in the observer test: https://github.com/apache/flink-kubernetes-operator/pull/26/files#diff-ac210d80370718839b3bdfc124a9dfc2ed32b06e35ca71474efd5a9f27203595R92-R116

wangyang0918 replied:
You are right.

bgeng777 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/BaseReconciler.java:
```diff
@@ -54,81 +47,22 @@ public BaseReconciler(KubernetesClient kubernetesClient, FlinkService flinkServi
             shutdown(flinkApp, effectiveConfig);
         } else {
             FlinkUtils.deleteCluster(flinkApp, kubernetesClient, true);
         }
```
Should we also call `FlinkUtils.waitForClusterShutdown()` here? I find `FlinkUtils.waitForClusterShutdown()` is only used for session cluster. And In `waitForClusterShutdown`, I notice that it will wait until JM and other service to be cleared. Why `deleteCluster` is not enough? If it cannot fullfill our requirements, should we wait for other services to be cleared in the `else` branch here?

gyfora replied:
FlinkUtils.deleteCluster will call waitForClusterShutdown internally if deleteHaConfigmaps is set to true. This has been introduced recently to clean up properly on deletion

bgeng777 replied:
Got it. I miss the use in the `waitForClusterShutdown` and it should work fine as we set the `deleteHaConfigmaps` to be true here. It seems that we can refine `stopSessionCluster` to avoid unnecessary `waitForClusterShutdown` when `deleteHaConfigmaps` is true. But it needn't be included in this PR. This PR LGTM. Thanks a lot for the work!

bgeng777 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/FlinkDeploymentStatus.java:
```diff
@@ -26,6 +28,8 @@
 @AllArgsConstructor
 public class FlinkDeploymentStatus {
-    private JobStatus jobStatus;
+    private JobStatus jobStatus = new JobStatus();
```
Not a crucial one: having default value for `jobStatus` field or not both make sense to me. But I am a little curious if there is any specific reason for adding it here. If there is, maybe we can refactor `createJobStatus()` in Observer.java to a static method here like `parseFromJobMessage()`. Whatever, this one is not critical for this PR and we can leave it as a followup if we hope.

gyfora replied:
Having a default a jobstatus avoids a bunch of null checks, and also not that you mention it I simplified the status update in a commit: https://github.com/apache/flink-kubernetes-operator/pull/26/commits/d4924c0849976930d591a6f60ab77759a4582e5b

SteNicholas commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java:
```diff
@@ -0,0 +1,162 @@
+    private void updateJobStatus(JobStatus status, List<JobStatusMessage> clusterJobStatuses) {
+        Collections.sort(
+                clusterJobStatuses,
+                (j1, j2) -> -1 * Long.compare(j1.getStartTime(), j2.getStartTime()));
```
```suggestion
                (j1, j2) -> Long.compare(j2.getStartTime(), j1.getStartTime()));
```

SteNicholas replied:
@gyfora, this change affects the failure of the check style.

SteNicholas commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java:
```diff
@@ -0,0 +1,162 @@
+            clusterJobStatuses = flinkService.listJobs(effectiveConfig);
+        } catch (Exception e) {
+            LOG.error("Exception while listing jobs", e);
+            flinkAppStatus.getJobStatus().setState("UNKNOWN");
```
I'm confused that whether to set the state to `UNKNOWN` when listing jobs exception. IMO, the listing jobs exception is same as no jobs found, therefore this doesn't update the job status.

gyfora replied:
I have added this so I can simply call observe when we delete the flink resource and only try to cancel if the job is RUNNING.

UNKNOWN does not say that it is running/failing it barely means that the operator cannot access it so flink client operations are not possible. It might not be perfect but I think it makes sense

SteNicholas reviewed (approved):
LGTM.

tweise commented:
@gyfora with the latest from this PR, I don't see the deployment go into `READY` state. It just loops with `JobManager deployment basic-example in namespace default port ready, waiting for the REST API...`

tweise commented:
After manually deleting the old CRD (helm uninstall did not do it) the above issue goes away.
